File: frontend/src/components/transactions/TransactionEditor/accountMove.ts
/**
 * What the editor's Account field may do on a row that already exists.
 *
 * The rule itself lives on the server — `backend/src/igab/domain/account_move.py`
 * refuses a move it must refuse, whatever any client believes. This module
 * decides only whether to *offer* the control, and what to say instead when it
 * does not: an editor that shows an enabled picker for a bank-fed row teaches
 * the user to expect a move and then hands them a 400.
 *
 * So the two copies answer different questions and cannot silently disagree
 * about money: the worst a stale copy here can do is offer a control the
 * server then declines, in its own words.
 *
 * Kept pure (facts in, string out) so every branch is a one-line test.
 */

import { rowMayCarryCategory } from '../../../utils/rowCategoryRule'

export interface AccountLockFacts {
  /** The statement vouched for a balance in one account. */
  isReconciled: boolean
  /** The row came from a bank feed (`sync_id`), which owns which account it
   *  lives in — the feed is account-scoped and would re-add it. */
  isBankFed: boolean
  /** A split's line. Its account is the parent's; the parent is the editable
   *  row, and moving it takes the lines along. */
  isSplitLine: boolean
}

/**
 * Why this row's account cannot be changed here, or null when it can.
 *
 * The sentence is shown beside the field, so it says what to do next rather
 * than only what is forbidden.
 */
export function accountLockReason(facts: AccountLockFacts): string | null {
  if (facts.isSplitLine) {
    return 'Split lines live in the same account as the transaction they belong to.'
  }
  if (facts.isReconciled) {
    return 'Reconciled — a statement vouched for this account. Unlock from the row menu to move it.'
  }
  if (facts.isBankFed) {
    return 'Your bank feed decides which account this one lives in. Delete it and enter it by hand to put it elsewhere.'
  }
  return null
}

/**
 * The note under the picker when the account the row is heading for cannot
 * hold the category it currently has. Saving clears it — say so before the
 * save, not after.
 *
 * Asks `rowMayCarryCategory` rather than re-spelling it: this is the same
 * rule the field's visibility already uses, pointed at the target account.
 */
export function categoryDropNote(
  targetOnBudget: boolean,
  categoryName: string | null
): string | null {
  if (rowMayCarryCategory(targetOnBudget) || !categoryName) return null
  return `Tracking accounts don't carry categories — ${categoryName} will be cleared.`
}

File: backend/src/igab/domain/account_move.py
"""Whether a transaction may move to a different account.

A row's account is the field three other systems believe they own:

- **the bank feed**, which reported the row in one account and looks for it
  there again (`find_by_sync_id` is account-scoped);
- **reconciliation**, which vouched for a balance in one account — so
  `account_id` is in `RECONCILED_LOCKED_FIELDS` beside amount and date;
- **the transfer pair**, whose legs carry payees naming each other's
  accounts, so moving one leg renames the other's payee.

Only the second was written down before a move was possible at all. This
module is the other two, plus the structural rules — as facts in, message
out, so every branch is a one-line test with no database.

What it does NOT decide, because one implementation each:

- whether the resulting row may keep its category — `leg_may_carry_category`
  in `domain.transfers` answers that, and the service asks it about the
  account the edit *ends* on rather than the one it started on;
- whether a reconciled row may move — `domain.reconciliation` already locks
  `account_id`, and the service's generic locked-field check runs first.
"""

import uuid
from dataclasses import dataclass


@dataclass(frozen=True)
class MoveRequest:
    """A proposed move, as facts rather than ORM rows.

    `counterpart_account_id` is the far leg's account when this row is a
    linked transfer leg, and None otherwise.
    """

    target_account_id: uuid.UUID
    target_is_closed: bool
    is_split_child: bool
    is_bank_synced: bool
    counterpart_account_id: uuid.UUID | None = None
    retargeting_transfer: bool = False


def refusal_for_move(request: MoveRequest) -> str | None:
    """Why this move must not happen, or None when it may.

    One sentence, addressed to the person who asked for it: every one of
    these is reachable from the editor's account picker, and "400 Bad
    Request" is not an answer about their money.
    """
    if request.is_split_child:
        return (
            "Move the split's parent transaction instead — a split's lines "
            "always live in the same account as the row they belong to"
        )
    if request.is_bank_synced:
        return (
            "This transaction came from a bank feed, which decides which account "
            "it lives in. Moving it would leave the feed free to add it back to "
            "the original account, and you would have it twice. Delete it here "
            "and enter it by hand in the right account instead"
        )
    if request.target_is_closed:
        return "That account is closed — reopen it first if this transaction belongs there"
    if request.retargeting_transfer:
        return (
            "Change the account this transaction is in and the account it "
            "transfers to in separate edits"
        )
    if request.counterpart_account_id is not None and (
        request.counterpart_account_id == request.target_account_id
    ):
        return (
            "This is a transfer to that account, so both sides would end up in it. "
            "Break the transfer first, or send it somewhere else"
        )
    return None
